### PR TITLE
[FLINK-4988] Elasticsearch 5.x support

### PR DESCRIPTION
**Pull request by mikedias**
Provides Elasticsearch 5.x support based on previous 2.x codebase.

### CODE REVIEW

**Comment by StephanEwen** [comment]
Is elastic search backwards compatible? Is there a way to support multiple Elasticsearch versions in one connector?
Would be nice not having to maintain three different versions of the ElasticSearch connector.

@rmetzger What is your take on this?

**Comment by StephanEwen** [comment]
From taking a quick look at it, I would suggest that this connector shades netty away. That way we can avoid conflicts whenever we adjust Flink's internal netty version.

**Comment by mikedias** [comment]
No, ES is not backward compatible... But we can reuse some classes or interfaces between versions. I have plans to do this in another PR, just for isolate possible issues.

**Comment by mikedias** [comment]
Not sure if I can exclude netty4 dependency, but I'll take a look.

**Comment by StephanEwen** [comment]
If you cannot exclude the netty dependency, you could try to shade it away: https://maven.apache.org/plugins/maven-shade-plugin/

We do this all the time for conflicting dependencies.

**Comment by StephanEwen** [comment]
@mikedias @rmetzger How do we proceed with this pull request?
Should it become part of Flink, Bahir, or should hosted in the contributor's repository?

File: flink-streaming-connectors/flink-connector-elasticsearch5/pom.xml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+		<dependency>
+			<groupId>org.apache.logging.log4j</groupId>
+			<artifactId>log4j-api</artifactId>
+			<version>2.7</version>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.logging.log4j</groupId>
+			<artifactId>log4j-core</artifactId>
+			<version>2.7</version>
```
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-elasticsearch5/pom.xml]: Why did you add log4j2 dependencies to the project?

File: flink-streaming-connectors/flink-connector-elasticsearch5/pom.xml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+			xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+			xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>org.apache.flink</groupId>
+		<artifactId>flink-streaming-connectors</artifactId>
```
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-elasticsearch5/pom.xml]: The parent's name has changed in the meantime (we've refactored our maven structure a bit for the connectors)

**Comment by rmetzger** [comment]
@StephanEwen I'll start a discussion on the mailing list to decide how we want to proceed.

**Comment by mikedias** [comment]
I have no problem in hosting the connector in my github account.

File: flink-streaming-connectors/flink-connector-elasticsearch5/pom.xml (hunk 1 above)
Comment (mikedias) [inline flink-streaming-connectors/flink-connector-elasticsearch5/pom.xml]: Elasticsearch needs a log4j2 implementation in classpath: https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking_50_java_api_changes.html#_elasticsearch_will_no_longer_detect_logging_implementations

**Comment by tzulitai** [comment]
@mikedias @rmetzger @StephanEwen 
I've also responded to the discussion in the ML started by Robert with a +1.

A recap of the proposed approach on how we proceed:
since the ES connectors share a lot of code, we'll refactor the ES connectors by having a `flink-connector-elasticsearch-base` module, and let version specific connections stem from that; basically how we're currently maintaining the Kafka connectors. There are two +1 votes for this approach, and no other objections, so I think we can agree on proceeding.

@mikedias, are you be ok with me using your PR as a base to refactor the ES connectors? I'll open a new PR with your changes and mine together. I'll also let you know when the PR is opened :)

**Comment by mikedias** [comment]
@tzulitai sure, no problem! :)

File: flink-connectors/flink-connector-elasticsearch5/src/main/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchSink.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+	/**
+	 * Creates a new ElasticsearchSink that connects to the cluster using a TransportClient.
+	 *
+	 * @param esConfig                  The map of user settings that are passed when constructing the TransportClient
+	 * @param sinkConfig                The map of user settings that are passed when constructing the BulkProcessor
+	 * @param transportAddresses        The Elasticsearch Nodes to which to connect using a {@code TransportClient}
+	 * @param elasticsearchSinkFunction This is used to generate the ActionRequest from the incoming element
+	 */
+	public ElasticsearchSink(Map<String, String> esConfig, Map<String, String> sinkConfig, List<InetSocketAddress> transportAddresses, ElasticsearchSinkFunction<T> elasticsearchSinkFunction) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-elasticsearch5/src/main/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchSink.java]: Hi @mikedias,
While refactoring the ES connectors based on your PR, I noticed that you changed the constructor's arguments to take the config for the client & bulk processor as separate `Map`s.

Is there an absolute need for this? Let me know if I'm missing anything, otherwise I'd prefer to keep this as consistent as possible across the different ES versions (I'll change this as part of the new PR, if you agree).

**Comment by rmetzger** [comment]
Sounds good. Thank you for taking care of this @tzulitai

File: flink-connectors/flink-connector-elasticsearch5/src/main/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchSink.java (hunk 3 above)
Comment (mikedias) [inline flink-connectors/flink-connector-elasticsearch5/src/main/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchSink.java]: Hi @tzulitai, It is necessary because ES 5.x is strict about configuration, not accepting any extra configuration that is not recognized. See https://www.elastic.co/guide/en/elasticsearch/reference/5.x/breaking_50_settings_changes.html

Comment (tzulitai) [inline flink-connectors/flink-connector-elasticsearch5/src/main/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchSink.java]: I see, thanks for the explanation! I think we can resolve this be keeping a single Map for user configuration at the API level, and internally, we separate out the bulk processor settings.

**Comment by tzulitai** [comment]
Hi @mikedias! Your contribution has been merged with 8699b03d79a441ca33d9f62b96490d29a0efaf44 and b5caaef82add4a6f424094d526700c77b011724e. Could you manually close this PR? The bot only closed the new restructure PR.

Thanks a lot for your contribution!

**Comment by rmetzger** [comment]
I closed the PR manually using another commit.
